recommendations.md

```diff
@@ -60,3 +60,204 @@ On the catalog, this means that you should, where possible:
         runAsUser: 0 # root uid == 0
         privileged: true
   ```
+
+## Don't use interpolation in shell scripts
+
+Using `$(tekton.task)` interpolation in the `script` or as a `sh -c`
+string is extremely fragile.  The interpolation done by tekton is not
+aware of the context in which the interpolation happens.  A space, a
+quote sign, a backslash or newline could easily thwart an otherwise
+beautiful script.
+
+```
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.one)
+```
+
+If `params.one` happens to contain a quote, then the resultin shell
+script might look like this:
+
+```
+echo '
+```
+
+This script is not valid, and the task will fail.
+
+Instead, use environment variables or arguments, which are not
+interpolated into the script
+
+```
+  steps:
+  - name: foo
+    image: myimage
+    env:
+      - name: PARAM_ONE
+        value: $(params.one)
+    script: |
+      echo "$PARAM_ONE"
+```
+
+The script will now correctly print out the value of $PARAM_ONE,
+regardless of what it contains.
+
+
+## Extract task code (scripts) to their own files
+
+As a task grows in complexity, it becomes harder and harder to
+maintain it in-line.  Because you have already avoided interpolation
+in the script, there is no real need for the script to be in-lined
+into the Task.
+
+Use a ConfigMap for your script, and mount the configmap in the task.
+Use a single _command:_ that executes your script (optionally with
+environment variables and with parameters):
+
+
+```
+  volumes:
+  - name: scripts
+    configMap:
+      name: my-task-scripts
+      defaultMode: 0755
+  steps:
+  - name: foo
+    image: myimage
+    volumeMounts:
+    - name: scripts
+      mountPath: /mnt/scripts
+    volume
+    command: 
+    - /mnt/scripts/my-command.sh
+```
+
+This allows you to create and update your configmap from a real
+script.  An external script-file allows you to edit it stand-alone as
+a proper script file, rather than in-lined into a tekton task yaml
+file.  This gives you better completion, syntax checking, and many
+other benefits.  You can even run the script locally.
+
+
+```
+#!/bin/bash
+# my-command.sh
+echo "${PARAM_ONE-Hello world}"
+```
+
+To create the configmap:
+
+```
+kubectl create configmap my-task-scripts --from-file=my-command.sh
+```
+
+For bonus points, use a kustomize generator to create your task; this
+allows you simple "kubectl apply -k" 
+
+## Remember that there are other languages than sh and bash
+
+Yes, sh and bash are DSLs for running processes, but sometimes there
+are other languages more suited for what you're trying to do.  Tekton
+Pipelines' main positive attribute is the ability to have the right
+tool available for every step, and that includes the _interpreter_.
+Use python or other scripting languages when that is warranted.
```

Comment:
FYI: this is how we do it in Python : 

https://github.com/tektoncd/catalog/blob/master/task/github-add-comment/0.2/github-add-comment.yaml#L76-L78

It has the same issue as shell, if you use `var = "$(params.foo)"` and params.foo has a " then you run into problems

Reply:
Yes, I'll 
* add a python example of the escaping issue
* add a pyton example on how to invoke a python script!